### DISCUSSION TITLE
Considering turning `Svelte-Embla` as an offically supported embla package

### DISCUSSION THREAD

**Discussion by Yofou** [davidjerleke/embla-carousel#333] [Ideas]
# The Idea
as you may know, a while back I mentioned I had made a Embla adapters for Svelte earlier this year called [svelte-embla](https://svelte-embla.vercel.app/) in  #307 , after some significant life changes between the period of when I posted that issue and now I've been quite busy adjusting to them. however now I have finally settled down and ready to commit some time to discuss about turning my svelte adapter for embla into an official supported package.

things to note, this is a really small adapter, the adapter it's self fit's nicely in [one file here](https://github.com/Yofou/svelte-embla/blob/main/src/lib/index.ts), which should making the task if so to happen of moving it to the official repository a breeze.

feel free to go explore the package and tell me what you think, maybe there should be some improvement that need to be made before moving it over

**Comment by davidjerleke** [comment]
Hi @Yofou,

Thank you for your work so far. It's highly appreciated 👍. Your Svelte wrapper is well aligned with the Embla Carousel vision: A bare bones library that does the heavy lifting and enables users to extend it to their liking with its highly flexible API. I'm going to suggest a few additions and small adjustments that I hope will make sense to you:

- Only dispatch an `init` event and let users hook on to other Embla events by exposing the carousel API. The main advantage with this approach is that we don't need to re-define Embla events in the Svelte wrapper. This won't cause the Svelte wrapper to get out of sync with the core package if events were to change in the core package in the future ([code](https://codesandbox.io/s/jolly-glade-4k7h8o?file=/emblaCarouselSvelte.js:366-736)).
- The action update method should re-initialize the carousel if options or plugins were to change. But, the update method should only re-initialize Embla if options or plugins actually change, and not on every reassignment ([code](https://codesandbox.io/s/jolly-glade-4k7h8o?file=/emblaCarouselSvelte.js:797-1393)).

Here's a simple [CodeSandbox](https://codesandbox.io/s/jolly-glade-4k7h8o?file=/emblaCarouselSvelte.js:797-1393) that demonstrates the changes above.

### 🚨 **Note!**

I didn't have much time to create my CodeSandbox example so it's not using TypeScript. But the Svelte package will be written in TypeScript once we agree on its contents.

Let me know what you think and thank you for your time!

Cheers,
David

**Reply by Yofou** [reply]
Hi @davidjerleke 

I really appreciate you checkout out the library and looking for things to improves, It looks like things are getting off at a great start 😄 .

## "Only dispatch an init event"
so let's start off with your first point about just dispatching init, I got alot to say about this so prepare yourself haha, I agree I think we should try solving the events from ever going out of sync as much as possible from version to version, however I think we should maybe look at what we can add/change in the base Emba-carousel code to help with keeping the wrapper(s) in sync before considering removing features from the wrapper. so I spent my sunday morning and bit of my afternoon (I got really invested haha) and I came up with this plan and I got a working prototype [here](https://github.com/Yofou/svelte-embla/blob/propose/embla-offical-changes/src/lib/index.ts). so if you flick through you'll notice there's a [`emblaEventType`](https://github.com/Yofou/svelte-embla/blob/36e0aa5949ebdbb8024363c8cf545adf1cbc3660/src/lib/utils.ts#L4-L14) which is the bread and butter of this idea. which we can iterate and dispatch automatically and... I beilive this idea can fit into the internal `Embla-carousel` lib quite nicely [here](https://github.com/davidjerleke/embla-carousel/blob/7aa1b3d658a768e42999622d49ccd7eb4e9fd78e/packages/embla-carousel/src/components/EventHandler.ts#L4-L13).

so rather than having
```ts
export type EmblaEventType =
  | 'init'
  | 'pointerDown'
  | 'pointerUp'
  | 'scroll'
  | 'select'
  | 'settle'
  | 'destroy'
  | 'reInit'
  | 'resize'
```

we actually have two exports like this
```ts
export const emblaEventType = [
	"init",
	"pointerDown",
	"pointerUp",
	"scroll",
	"select",
	"settle",
	"destroy",
	"reInit",
	"resize",
] as const;

export type EmblaEventType = typeof emblaEventType[number]  // this gets the union literal types as before (ie  "init" | "pointerDown" | "pointerUp" | ...)
```
this keeps the code dry & type safe and let's me (or anyone else making an embla wrapper in the future) have the option of automatically dispatching events while not having to worry about maintaining there own derived list.

I think this creates a best of both worlds, let me know what you think.

also another bullet point to make is if you go to [Svelte-Embla docs](https://svelte-embla.vercel.app/guides/stores) right now you'll notice there's a svelte writable-store (which if you don't know is a built-in supported observable in svelte it'self) that you can pass in currently to cut out the middle man of grabbing the embla client through the init event, Which I believe is alot nicer & the more "sveltier" way of going around it, but I still believe we should have the init event details pass the client though to not hinder anyone's options on how they would like to grab the embla client.

## "The action update method should re-initialize the carousel if options or plugins were to change"
100%, that is a really good catch which I have already added to the [proposal branch](https://github.com/Yofou/svelte-embla/tree/propose/embla-offical-changes) on my repo with the typescript types added for you 👍🏻 

# 📔 Note
In that proposal branch I also experimented with the idea of also automatically telling svelte (LSP) to add type completion for the embla events on the element, which went some-what successful, there is no way for me to inform typescript to only give the type completion if the embla action has been added to the element it'self which means those embla events alway popup & I can't seem to get jsdocs comments to show up to inform the user either to add the action, which is little bit of a bummer. but it is very convenient to have these type annotations pop up while developing (futher more supporting why I think we should keep them).

![image](https://user-images.githubusercontent.com/39814328/178148376-fd63702e-d501-45d4-bdc4-598506e261ff.png)

anyway, I hope you have a great day & that I haven't given you to much eye strain from long comment. 👋

**Reply by davidjerleke** [reply]
Hi @Yofou,

With all there is to do with the upcoming v7, the time to response is not quite what I would like it to be. Thank you for your patience 🙂.

## Events

> however I think we should maybe look at what we can add/change in the base Emba-carousel code to help with keeping the wrapper(s) in sync before considering removing features from the wrapper.

I don't agree with this. Mainly because if we were to change the Embla core for every framework wrapper that was introduced, it will make a mess out of the core in the long run. I think the main goal with a framework wrapper should be to get rid of the main friction using the library together with a framework. It shouldn't add duplicate layers of logic and a middle hand, but rather act as a bridge and expose the native library API as is. There are important benefits with this approach:

- No need for duplicate logic which will prevent the wrapper and library to get out of sync. You mentioned that you had busy a period in your life and I have it too sometimes. We want to avoid doing unnecessary work. 

- Types will work as expected because Embla, its wrappers and plugins are all written in TypeScript. As you mentioned in your [response](https://github.com/davidjerleke/embla-carousel/discussions/333#discussioncomment-3115767) it wasn't straightforward to add event types to the element.

You can easily use the Embla API in the [CodeSandbox](https://codesandbox.io/s/jolly-glade-4k7h8o) I provided and adding event hooks will work great with autocomplete:

<img width="558" alt="events" src="https://user-images.githubusercontent.com/11529148/179602671-10c3f6ec-5c98-4830-be1a-5b929c3e2739.png">

A limitation with adding a duplicate event layer to the Svelte wrapper is that Embla allows for adding multiple event listeners that listens to the same event:

```ts
// Add event listeners individually
embla.on('select', togglePrevNextButtonsActive)
embla.on('select', toggleDotButtonActive)
```

And it also gives you full control over removing events individually:

```ts
embla.off('select', toggleDotButtonActive) // Remove event listener toggleDotButtonActive, but togglePrevNextButtonsActive will remain active
```

This feature would be lost when creating the Svelte event layer.

It's important that the Embla events remain a type definition and not an actual JavaScript object because plugins will extend the event list type when a TypeScript user installs and uses plugins. Adding events to the event object every time a plugin is initialized and removing them when the plugin is destroyed will get messy. 

Additionally, your suggestion would result in TypeScript emitting actual code, an object literal for the events, which will increase the bundle size. People stare themselves blind on bundle sizes these days so I have to weigh that in every time I make changes to the core.

The [embla-carousel-react](https://github.com/davidjerleke/embla-carousel/tree/master/packages/embla-carousel-react) wrapper has been around for some years and it **exposes the raw Embla API**. So speaking from experience, it has barely needed any maintenance really. It just works. And I just finished the [embla-carousel-vue](https://github.com/davidjerleke/embla-carousel/tree/master/packages/embla-carousel-vue) wrapper the other day which also successfully exposes the **raw Embla API**. Both these wrappers will automatically adapt to any changes made to the Embla API without the need of manual labour. 

## Store

> also another bullet point to make is if you go to [Svelte-Embla docs](https://svelte-embla.vercel.app/guides/stores) right now you'll notice there's a svelte writable-store (which if you don't know is a built-in supported observable in svelte it'self) that you can pass in currently to cut out the middle man of grabbing the embla client through the init event, Which I believe is alot nicer & the more "sveltier" way of going around it, but I still believe we should have the init event details pass the client though to not hinder anyone's options on how they would like to grab the embla client.

I'm not very experienced with Svelte. So I dived into this MDN guide hoping to understand stores better:

> Moreover, when your app becomes complicated and your component hierarchy gets complex, it might become too difficult for components to relay data between each other. In that case, moving to a global data store might be a good option. If you've already worked with [Redux](https://redux.js.org/) or [Vuex](https://vuex.vuejs.org/), then you'll be familiar with how this kind of store works. Svelte stores offer similar features for state management.

From my understanding, a store is basically a state management feature in Svelte. For a carousel component, it seems a bit overkill because in most cases the carousel instance will be local to the component and not used across components. At least it has been for me when I think about how I use it. Having said that, I understand that there might be cases where one would want the carousel instance to be accessible elsewhere.

But as I said I'm not very experienced with Svelte stores so I don't know the implications of a store, like how it affects performance and so on. If a user still has the possibility to assign their carousel variable with `event.detail` even when using the store, I guess it could be nice to offer users both choices. So I think your suggestion makes sense if this is the case.

Thank you for your time and let me know what you think.

Cheers,
David

**Reply by davidjerleke** [reply]
Hi @Yofou,

Not to stress you or anything but have you had the chance to read my [response](https://github.com/davidjerleke/embla-carousel/discussions/333#discussioncomment-3174677) yet? I’m close to a stable v7 release so this will soon be the only thing left to do.

Best,
David

**Reply by davidjerleke** [reply]
Hi @Yofou,

Maybe you're having your vacation or maybe you're very busy. I couldn't wait for the v7 release any longer so I created a simple Svelte wrapper that was published with the release. I'm here if you want to continue this conversation and still waiting for a response on my [last comment](https://github.com/davidjerleke/embla-carousel/discussions/333#discussioncomment-3174677). No need to hurry, just wanted to be clear.

Best,
David

**Comment by ecker00** [comment]
The store approach would make this much more usable. 👍 I tried both the original package vs the one integrated now, and would not use this one in it's current state. I got both of them working, but the current approach throws TypeScript warnings when you put `<div on:init={…}>`, as div's don't have this signature.

Also the store approach would making porting the documentation over much easier. Currently there is very little usable documentation and took a lot of experimentation to get it working.

**Reply by davidjerleke** [reply]
Hi @ecker00,

Thank you for your input.

> Currently there is very little usable documentation and took a lot of experimentation to get it working.

Yes, unfortunately the Svelte and Vue packages don't have much of a documentation yet. I'm the sole developer working on this project so it's hard to find the time needed to improve that. 

> would not use this one in it's current state.

I helped a friend of mine getting started with `embla-carousel-svelte` the other day. He had no problems using the package and the Embla API to build a carousel with arrows, dots etc. What problems do you run in to?

> I got both of them working, but the current approach throws TypeScript warnings when you put `<div on:init={…}>`, as div's don't have this signature.

Did you run into the TypeScript problem with both packages or only `embla-carousel-svelte`? Because both attach custom event listeners to a div. One solution is to add the following to your `d.ts` file:

```ts
declare namespace svelte.JSX {
    interface HTMLProps<T> {
        oninit?: (e: CustomEvent) => void;
    }
}
```

I don't think this is clean way to do it though as it will accept the `init` event on any HTML element. Do you know a better way to do this? If yes, please share.

> The store approach would make this much more usable. 👍

I'm not an experienced Svelte dev at all, more of a novice because my job is 100% React and has been for the last 4 years. But the conversation in this discussion so far has pretty much been: A store is better. It's more of a Svelte way to do things. But I never got an answer to **why** it's better. Please, don't get me wrong, I'm open to suggestions and improvements but I just want a good reason for it. In [this comment](https://github.com/davidjerleke/embla-carousel/discussions/333#discussioncomment-3174677), you can see my previous response about the store thing: 

<div type='discussions-op-text'>

<sup>Originally posted by **davidjerleke** July 18, 2022</sup>

I'm not very experienced with Svelte. So I dived into this MDN guide hoping to understand stores better:

<br />

> Moreover, when your app becomes complicated and your component hierarchy gets complex, it might become too difficult for components to relay data between each other. In that case, moving to a global data store might be a good option. If you've already worked with [Redux](https://redux.js.org/) or [Vuex](https://vuex.vuejs.org/), then you'll be familiar with how this kind of store works. Svelte stores offer similar features for state management.

From my understanding, a store is basically a state management feature in Svelte. For a carousel component, it seems a bit overkill because in most cases the carousel instance will be local to the component and not used across components. At least it has been for me when I think about how I use it. Having said that, I understand that there might be cases where one would want the carousel instance to be accessible elsewhere.

<br />

But as I said I'm not very experienced with Svelte stores so I don't know the implications of a store, like how it affects performance and so on. If a user still has the possibility to assign their carousel variable with `event.detail` even when using the store, I guess it could be nice to offer users both choices. So I think your suggestion makes sense if this is the case.

</div>

<br />

Please share your view about this and if I'm missing anything important here, and why it's beneficial to use a Svelte store internally in the `embla-carousel-svelte` package. Wouldn't it be possible for devs to still use a store if they wanted to? Something along these lines:

```svelte
<script>
  import emblaCarouselSvelte from 'embla-carousel-svelte'
  import { writable } from 'svelte/store'

  let emblaApiStore = writable

  const emblaConfig = {
    options: { loop: false },
    plugins: []
  }

  const onInit = (event) => {
    emblaApiStore.set(event.detail)
    // Embla API is ready
  }
</script>

<div class="embla" use:emblaCarouselSvelte={emblaConfig} on:init={onInit}>
  <div class="embla__container">
    <div class="embla__slide">Slide 1</div>
    <div class="embla__slide">Slide 2</div>
    <div class="embla__slide">Slide 3</div>
  </div>
</div>
```

Thank you for your time.

Best,
David

**Reply by ecker00** [reply]
> What problems do you run in to?

Got it working in the end (just guessing the documentation), but was the typing issue which you do suggest a solution to, but saying HTML props have `oninit` is kind of just a band-aid. 😅

> Store …

I can't speak for under the hood details. But stores are very developer friendly and removes a lot of boilerplate code. I've used them a lot in a current application and see no performance impacts.

For us what we wanted is to just have a custom component we can easily interact with. We're in a situation we need to ship fast atm, and just needed an easy approach that felt familiar. Thus we ended up with `https://github.com/vadimkorr/svelte-carousel` for now. This is using a custom component which exposes it's inner API, very little boilerplate code.
```svelte
<Carousel bind:this={carousel}>
```

Maybe I'm a bit quick to conclude on store. The important bit imo is the custom component, and this could probably be achieved without using a store.

**Reply by davidjerleke** [reply]
Hi @ecker00,

Thanks for the additional details.

> but was the typing issue which you do suggest a solution to, but saying HTML props have oninit is kind of just a band-aid. 

Yes. As I mentioned I don't think it's a clean solution. But Svelte should have a way of attaching custom events to any element right? What's the "good" way then 🙂?

> This is using a custom component which exposes it's inner API, very little boilerplate code.

Sure. If I understand you correctly, you want a carousel with most features built-in. If that's the case, there are lots of libraries out there. However, Embla doesn't aim to be that, but rather this (from the [docs](https://www.embla-carousel.com/get-started/)):

> Embla Carousel is a library agnostic, dependency free and lightweight carousel library. It aims to solve the hardest technical challenges with building carousels, and the rest is up to the user utilizing its highly extensible [API](https://www.embla-carousel.com/api/) and [plugins](https://www.embla-carousel.com/plugins/). 

...so Embla is more of a **DIY** library. It doesn't aim to be a library with an exhaustive set of features covering most use cases. It should basically solve the hardest technical challenges with building great carousels and do it good. The rest of the implementation should be up to the user, by utilizing its [extensible API](https://www.embla-carousel.com/api/) and [plugins](https://www.embla-carousel.com/plugins/). 

Best,
David

**Reply by ecker00** [reply]
Don't need to do anything more than it does today, just how the basic functions are exposed would make it more friendly to integrate and interact with. 

There are multiple ways to approach it, and maybe a good example would be the Svelte-Embla that already exist:
https://svelte-embla.vercel.app/guides/options

Another way would be something like this (I'm just writing from top of my head):
```svelte
<script>
  import Embla, { createEmbla } from 'embla';
  const embla = createEmbla({ /* options */ });
  
  // Now possible to interact with the API like:
  setTimeout(() => {
    embla.scrollNext();
  }, 1000);
</script>
<div>
  <Embla {embla}>
    <!-- Content -->
  </Embla>
</div>
```

**Reply by davidjerleke** [reply]
Seems like I'm misunderstanding you @ecker00. The current way of accessing the Embla API is like so:

```html
<script lang="ts">
  import emblaCarouselSvelte, { EmblaCarouselType } from 'embla-carousel-svelte'

  let emblaApi: EmblaCarouselType

  const onInit = (event) => (emblaApi = event.detail) // Assign Embla API when ready
  const scrollPrev = () => emblaApi?.scrollPrev() // Make use of the Embla API
  const scrollNext = () => emblaApi?.scrollNext() // Make use of the Embla API
</script>

<div class="embla" use:emblaCarouselSvelte on:init={onInit}>
  <div class="embla__container">
    <div class="embla__slide">Slide 1</div>
    <div class="embla__slide">Slide 2</div>
    <div class="embla__slide">Slide 3</div>
  </div>
</div>

<button on:click={scrollPrev}>Scroll prev</button>
<button on:click={scrollNext}>Scroll next</button>
``` 

And Svelte-Embla like so:

```html
<script lang="ts">
  import embla from "svelte-embla"
  import type { EmblaCarouselType } from "embla-carousel";
  import { writable } from "svelte/store";

  const emblaApi = writable<EmblaCarouselType>();
  const scrollPrev = () => $emblaApi?.scrollPrev(); // Make use of the Embla API
  const scrollNext = () => $emblaApi?.scrollNext(); // Make use of the Embla API
</script>

<div class="embla" use:embla={{ store: emblaApi }}> <!-- Assign Embla API when ready -->
  <div class="embla__container">
    <div class="embla__slide">Slide 1</div>
    <div class="embla__slide">Slide 2</div>
    <div class="embla__slide">Slide 3</div>
  </div>
</div>

<button on:click={scrollPrev}>Scroll prev</button>
<button on:click={scrollNext}>Scroll next</button>
```

What bothers you here is the assignment with the `onInit` event? Or am I missing other things you have in mind that aren't present in these examples (like what stores can do or similar)? I'm just looking to understand you better.

Thanks for your time.

Best,
David

**Reply by ecker00** [reply]
Yeah, you hit the nail on the head there. 👍 Sorry for making it a round about conversation. So many ways to achieve similar results, but a setup that's fully type safe and documented to work is plenty. Custom component was just an idea to encapsulate some of the default setup steps, not important.

**Reply by davidjerleke** [reply]
> Got it working in the end (just guessing the documentation), but was the typing issue which you do suggest a solution to, but saying HTML props have oninit is kind of just a band-aid. 

@ecker00 the problem with the types for the `init` **event** was just me missing to include it in `ActionReturn<Parameter, Attribute>`. This [pull request](https://github.com/davidjerleke/embla-carousel/pull/465/files#diff-790f6e048ece9fad3bad569c4294f6fe614ff236054831ca1ca7dbe48de47a9e) enables the tooling to pickup the correct types. So the following band aid in the `d.ts` file will **no longer be necessary**:

```ts
declare namespace svelte.JSX {
    interface HTMLProps<T> {
        oninit?: (e: CustomEvent) => void;
    }
}
```

Best,
David

**Reply by davidjerleke** [reply]
I've just released the bug fix with [v8.0.0-rc02](https://github.com/davidjerleke/embla-carousel/releases/tag/v8.0.0-rc02).
